File: c++/count-and-say.cpp
class Solution {
private:
    string intToStr(int n) {
        string s;
        while (n > 0) {
            s.push_back(n % 10 + '0');
            n /= 10;
        }
        reverse(s.begin(), s.end());
        return s;
    }

public:
    string countAndSay(int n) {
        /* https://oj.leetcode.com/problems/count-and-say/
        */
        
        if (n == 0) return "";
        
        string result = "1";
        for (int i = 1; i < n; i++) {
            string tmp;
            int j = 0;
            while (j < result.size()) {
                char c = result[j];
                int count = 0;
                while (j < result.size() && result[j] == c) {
                    count++;
                    j++;
                }
                tmp.append(intToStr(count));
                tmp.push_back(c);
            }
            result = tmp;
        }
        
        return result;
    }
};

File: python/count-and-say.py
class Solution:
    # @return a string
    def countAndSay(self, n):
        result = '1'

        for i in xrange(n-1):
            stat = self._count(result)
            result = self._say(stat)

        return result

    def _count(self, s):
        stat = []
        n = 0
        now = None
        for c in s:
            if now is None:
                now = c
                n = 1
            elif c == now:
                n += 1
            else:
                stat.append((now, n))
                now = c
                n = 1

        stat.append((now, n))
        return stat

    def _say(self, stat):
        return ''.join([str(n)+c for c, n in stat])

